src/Cooked/MockChain/Balancing.hs

```diff
@@ -1,409 +1,326 @@
+-- separated from the balancing.
+module Cooked.MockChain.Balancing (balanceTxSkel, getMinAndMaxFee, estimateTxSkelFee) where
 
+import Cardano.Api.Ledger qualified as Cardano
 import Cardano.Api.Shelley qualified as Cardano
-import Cardano.Ledger.Shelley.Core qualified as Shelley
 import Cardano.Node.Emulator.Internal.Node.Params qualified as Emulator
 import Cardano.Node.Emulator.Internal.Node.Validation qualified as Emulator
-import Control.Arrow
 import Control.Monad.Except
+import Cooked.Conversion
 import Cooked.MockChain.BlockChain
 import Cooked.MockChain.GenerateTx
+import Cooked.MockChain.MinAda
 import Cooked.MockChain.UtxoSearch
 import Cooked.Output
 import Cooked.Skeleton
 import Cooked.ValueUtils
 import Cooked.Wallet
+import Data.Bifunctor
 import Data.Function
 import Data.List
-import Data.Map (Map)
 import Data.Map qualified as Map
+import Data.Maybe
+import Data.Ratio qualified as Rat
 import Data.Set (Set)
 import Data.Set qualified as Set
-import Ledger.Index qualified as Ledger
-import Optics.Core hiding (chosen)
+import Optics.Core
 import Plutus.Script.Utils.Ada qualified as Script
-import Plutus.Script.Utils.Scripts qualified as Script
-import Plutus.Script.Utils.Value qualified as Script
+import PlutusLedgerApi.V1.Value qualified as Api
 import PlutusLedgerApi.V3 qualified as Api
-import PlutusTx.Numeric qualified as PlutusTx
+import PlutusTx.Prelude qualified as PlutusTx
 
-balanceTxSkel :: (MonadBlockChainBalancing m) => TxSkel -> m (TxSkel, Fee, Set Api.TxOutRef)
-balanceTxSkel skelUnbal = do
-  -- We retrieve the balancing wallet, who is central in the balancing
-  -- process. Any missing asset will be searched within its utxos.
-  let balancingWallet =
-        case txOptBalanceWallet . txSkelOpts $ skelUnbal of
-          BalanceWithFirstSigner -> case txSkelSigners skelUnbal of
-            [] -> error "Can't select balancing wallet: There has to be at least one wallet in txSkelSigners"
-            bw : _ -> bw
-          BalanceWith bWallet -> bWallet
+-- * A few types to make the functions in this module more readable
 
-  -- We collect collateral inputs. They might be directly provided in the
-  -- skeleton, or should be retrieved from a given wallet
-  collateralInputs <- case txOptCollateralUtxos . txSkelOpts $ skelUnbal of
-    CollateralUtxosFromBalancingWallet -> getCollateralInputs balancingWallet
-    CollateralUtxosFromWallet cWallet -> getCollateralInputs cWallet
-    CollateralUtxosFromSet utxos -> return utxos
+type Fee = Integer
 
-  -- We compute the balanced skeleton with the associated fees
-  (skel, fees) <-
-    if txOptBalance . txSkelOpts $ skelUnbal
-      then setFeeAndBalance balancingWallet skelUnbal collateralInputs
-      else return (skelUnbal, Fee 0)
+type Collaterals = Set Api.TxOutRef
 
-  -- We return the new skeleton, the fees and the collateral inputs
-  return (skel, fees, collateralInputs)
+type BalancingOutputs = [(Api.TxOutRef, Api.TxOut)]
```

Comment:
`Fee` used to be defined as a type in `Skeleton`, but I removed it. It did not make sense there. In the current state of this PR, `Fee` only exists as an alias in this specific module to improve redability, and is not visible by the rest of the project so I would advocate to keep it like this. I really dislike the additional sugar of the required constructor when using `newtype`

Reply:
Then instead of making it a newtype, we could export it to help make type signatures involving fees more self-explanatory outside this module? I am not 100% convinced. You seem to have stronger opinion about this so I trust your judgement.